utils/format_doc.mli

```diff
@@ -0,0 +1,281 @@
+  | Text of string
+  | With_size of int
+  | Open_box of { kind: box_type ; indent:int }
+  | Close_box
+  | Open_tag of Format.stag
+  | Close_tag
+  | Open_tbox
+  | Tab_break of { width : int; offset : int }
+  | Set_tab
+  | Close_tbox
+  | Simple_break of { spaces : int; indent : int }
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
+
+  | Deprecated of (Format.formatter -> unit)
+    (** Escape hatch: a {!Format} printer used to provide backward-compatibility
+        for user-defined printer (from the [#install_printer] toplevel directive
+        for instance). *)
+
+ (** Immutable document type*)
+type t
+type doc = t
+
+(** Empty document *)
+val empty: t
+
+(** [format ppf doc] sends the format instruction of [doc] to the Format's
+    formatter [doc]. *)
+val format: Format.formatter -> doc -> unit
+
+(** Fold over a document as a sequence of instructions *)
+val fold: ('acc -> element -> 'acc) -> 'acc -> doc -> 'acc
+
+
+module Core: sig
+  type ('a,'b) fmt = ('a, doc, doc,'b) format4
+
+  type printer0 = doc -> doc
+  type 'a printer = 'a -> printer0
+
+
+  (** {!msg} and {!kmsg} produce a document from a format string and its
+      argument *)
+  val msg: ('a,doc) fmt -> 'a
+  val kmsg: (doc -> 'b) -> ('a,'b) fmt -> 'a
+
+  (** {!printf} and {!kprintf} produce a printer from a format string and its
+      argument*)
+  val printf: ('a, printer0) fmt -> 'a
+  val kprintf: (doc -> 'b) -> ('a, doc -> 'b) fmt -> 'a
+
+  (** The functions below mirror {!Format} printers, without the [pp_print_]
+      prefix naming convention *)
+  val open_box: box_type -> int -> printer0
+  val close_box: printer0
+
+  val text: string printer
+  val string: string printer
+  val bytes: bytes printer
+  val with_size: int printer
+
+  val int: int printer
+  val float: float printer
+  val char: char printer
+  val bool: bool printer
+
+  val space: printer0
+  val cut: printer0
+  val break: spaces:int -> indent:int -> printer0
+
+  val custom_break:
+    fits:(string * int * string as 'a) -> breaks:'a -> printer0
+  val force_newline: printer0
+  val if_newline: printer0
+
+  val flush: printer0
+  val force_stop: printer0
+
+  val open_tbox: printer0
+  val set_tab: printer0
+  val tab: printer0
+  val tab_break: width:int -> offset:int -> printer0
+  val close_tbox: printer0
+
+  val open_tag: stag printer
+  val close_tag: printer0
+
+  val list: ?sep:printer0 -> 'a printer -> 'a list printer
+  val iter:
+    ?sep:printer0 -> iter:(('a -> unit) -> 'b -> unit) -> 'a printer
+    ->'b printer
+  val array: ?sep:printer0 -> 'a printer -> 'a array printer
+  val seq: ?sep:printer0 -> 'a printer -> 'a Seq.t printer
+
+  val option: ?none:printer0 -> 'a printer -> 'a option printer
+  val result: ok:'a printer -> error:'e printer -> ('a,'e) result printer
+  val either: left:'a printer -> right:'b printer -> ('a,'b) Either.t printer
+
+end
+
+(** {1 Compatibility API} *)
```

Comment:
We could have the functions that are not so nice and mostly for compatibility in a `Compat` submodule. But I am okay with the current approach as well, the functions in fact tend to make sense already.